README.md

## 🎯 Case JusBrasil - Full Stack Challenge
This project was executed based on the technical test for dev 1 on JusBrasil, which is in [this repository](https://github.com/bsilva0xbr/cross-experience-level-3-to-4-tech-test).

Basically, the main objective is to build a web service that displays the plans offered by the company. The customer receives three plan options, with two promotional packages. Then he is directed to the product's checkout page, where he adds his information and completes the purchase.
All information is validated on the front and backend, with object manipulation.

## ⚙️ Features
You can use the endpoint "**GET**Plans" to see which plans are available, and send your sign request by "**POST**SignPlan".




## 💻 How to run locally
    1. Clone this repo;
    2. In your terminal, go to repo directory;
    3. Execute $ npm i;
    4. Then, $ docker compose up;
    5. In your web browser, go to http://localhost:3000/;  

## 💻 How to test
    1. In your terminal execute: $ npm run test;

## 🚀 Stack

 - [Next.JS](https://nextjs.org/docs)
 
 **Libs**
 - uuid
 - Express
 - Jest
 - Axios
 - body-parser


## 🧾 Documentation
- [Workflow - Trello](https://trello.com/invite/b/x71LSkxL/ATTI59e4d662dce893b038689d8dc56dec6fE898F1C5/desafio-jus-brasil)
- [API - Postman](https://documenter.getpostman.com/view/19296902/2s8YYCv69S)

##  👩🏾 Author
- [GitHub](https://github.com/christenLeo)
- [LinkedIn](https://www.linkedin.com/in/leochristen/)




HISTORY.md
Please fill out this file with your decisions, opinions or improvements.

### Archtecture
- Created the src/ folder and moved pages/, components/, and resources/ into it
- Created server/ folder and inside it made the server.js where my endpoints are, and moved the data/ folder from the resources/ folder to the server/ folder
- Moved the available-cards.json file from the src/resources/cards/ to the server/data/ folder
- Created the _document.js file into the src/pages/ folder to add in it the stylesheets links to resolve a rendering bug, also used the Script tag from next/script at the bottom of the page to resolve the rendering bug as well
- Made the src/resources/constants/urls.js to store the base url path
  
### Back-End  
- Installed the Express framework to implement the API
- Created the GET plans endpoint using the url: localhost:3000/plans
- Created the POST New plan sign endpoint using the url: localhost:3000/signplan
- Created the requests.rest file to manually test the endpoints
- Changed the npm run dev and npm run start scripts in the package.json, in order that my app starts from the server not from the client
- Installed the uuid library to generate unique Ids
- Installed the body-parser library to cast the JSON info from the body to a common object to use it in the POST endpoint

### Front-End
- Created the src/components/individualCard/index.jsx -> component to render the Individual plans info
- Created the src/components/bundleCard/index.jsx -> component to render the Bundle plans info
- Installed the react-hook-form library in order to validate the inputs before send this info to the API
- installed the axios library to make easier do the API calls

### Tests  
- Changed the react version from React@17.0.2 to React@18.0.0 in order to install the testing library JEST
- Installed Jest and React testing library
- Made the tests/ folder and the index.test.jsx file to do the tests of the home page
- Added the test script in package.json
- Created and setted up the jest.config.js and the jest.setup.js files

tests/server.test.js
import dataBasePlansInfo from "../server/data/available-plans.json";
import dataBaseCardsInfo from "../server/data/available-cards.json";
import { v4 as idGenerator } from 'uuid';


const getPlansMock = () => {
    return dataBasePlansInfo.data.availablePlans;
};

const postSignPlanMock = (body) => {

    const {cardNum, month, year, cvv, ownerName} = body;

    if (!cardNum || !month || !year || !cvv || !ownerName) {
        return 'Please check your infos, all fields must be filled';
    }

    const cardOnlyNums = cardNum.split('').filter((char) => {return char !== " "});

    if (cardOnlyNums.length < 15 || cardOnlyNums.length > 16) {
        return 'Please check your card number';
    } 

    for (let i = 0; i < cardOnlyNums.length; i++) {
        let num = Number(cardOnlyNums[i]);

        if (String(num) === 'NaN') {
            return 'Please check your card number';
        }
    }

    const dbCard = dataBaseCardsInfo.data.cards.filter(card => card.number === cardNum);

    if (!dbCard[0]) {
        return "Card isn't registered";
    }

    if (dbCard[0].description === 'Just an expired card.') {
        return dbCard[0].description;
    }

    if (dbCard[0].description === 'A card with no balance available.') {
        return dbCard[0].description;
    }

    if (month < 1 || month > 12) {
        return 'Please check the month number';
    }

    const cvvArr = cvv.split('');

    if (cvvArr.length < 3 || cvvArr.length > 4) {
        return 'Please check your cvv number';
    }

    for (let i = 0; i < cvvArr.length; i++) {
        let num = Number(cvvArr[i]);

        if (String(num) === 'NaN') {
            return 'Please check your cvv number';
        }
    }


    const id = idGenerator();
    const newPlanSign = {
        id,
        cardNum,
        month,
        year,
        cvv,
        ownerName
    };

    return {message: 'Plan signed successfully', info: newPlanSign};
};    

describe('GET Plans', () => {
    it ('it return the plans', () => {
        const plans = getPlansMock();

        expect(plans).toBe(dataBasePlansInfo.data.availablePlans);
    });
});

describe('POST Sign Plans', () => {
    it('Happy road, all validations check', () => {
        const body = {
            cardNum: '5452 5389 4527 1089',
            month: '11',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const {message} = postSignPlanMock(body);

        expect(message).toBe('Plan signed successfully');
    });
    it('With one or more info missing', () => {
        const body = {
            cardNum: '',
            month: '11',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('Please check your infos, all fields must be filled');
    });
    it('With wrong credit card number, numbers missing', () => {
        const body = {
            cardNum: '5452 5389 4527',
            month: '11',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('Please check your card number');
    });
    it('With wrong credit card number, with letters in the middle', () => {
        const body = {
            cardNum: '5452 53f9 45t7 1089',
            month: '11',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('Please check your card number');
    });
    it('With wrong credit card number, card not registered', () => {
        const body = {
            cardNum: '5452 5349 4517 1035',
            month: '11',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe("Card isn't registered");
    });
    it('Trouble at card validation, card without balance', () => {
        const body = {
            cardNum: '4532 4957 4695 2291',
            month: '11',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('A card with no balance available.');
    });
    it('Trouble at card validation, card expired', () => {
        const body = {
            cardNum: '5361 9278 2540 8152',
            month: '11',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('Just an expired card.');
    });
    it('Other info errors, wrong month number', () => {
        const body = {
            cardNum: '5452 5389 4527 1089',
            month: '25',
            year: '2023',
            cvv: '123',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('Please check the month number');
    });
    it('Other info errors, wrong cvv number', () => {
        const body = {
            cardNum: '5452 5389 4527 1089',
            month: '11',
            year: '2023',
            cvv: '12342',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('Please check your cvv number');
    });
    it('Other info errors, wrong cvv number with letters', () => {
        const body = {
            cardNum: '5452 5389 4527 1089',
            month: '11',
            year: '2023',
            cvv: '1t3',
            ownerName: 'Leona Christen'
        }
        
        const res = postSignPlanMock(body);

        expect(res).toBe('Please check your cvv number');
    });
});

server/server.js
const express = require('express');
const next = require('next');
const bodyParser = require('body-parser');
const plans = require('./data/available-plans.json');
const cards = require('./data/available-cards.json');
const { v4: idGenerator } = require('uuid');
const dev = process.env.NODE_ENV !== 'production';

const app = next({dev});
const handle = app.getRequestHandler();
const jsonParser = bodyParser.json();

app.prepare().then(() => {
    const server = express();
    const port = 3000;

    // Endpoints
    server.get('/plans', (req, res) => {
        res.status(200).send(plans);
    });

    server.post('/signplan', jsonParser, (req, res) => {
        let statusCode = 200;
        try {
            const {cardNum, month, year, cvv, ownerName} = req.body;

            // Checking if the fields are filled
            if (!cardNum || !month || !year || !cvv || !ownerName) {
                statusCode = 400;
                throw new Error('Please check your infos, all fields must be filled');
            }

            // Checking if the card have 16 digits or 15 case the card is an American Express
            const cardOnlyNums = cardNum.split('').filter((char) => {return char !== " "});

            if (cardOnlyNums.length < 15 || cardOnlyNums.length > 16) {
                statusCode = 400;
                throw new Error('Please check your card number');
            } 

            // Checking if all the digits are numbers
            for (let i = 0; i < cardOnlyNums.length; i++) {
                let num = Number(cardOnlyNums[i]);

                if (String(num) === 'NaN') {
                    statusCode = 400;
                    throw new Error('Please check your card number');
                }
            }

            // Check if the card is registered
            const dbCard = cards.data.cards.filter(card => card.number === cardNum);

            if (!dbCard[0]) {
                statusCode = 401;
                throw new Error("Card isn't registered");
            }

            if (dbCard[0].description === 'Just an expired card.') {
                statusCode = 401;
                throw new Error(dbCard[0].description);
            }

            if (dbCard[0].description === 'A card with no balance available.') {
                statusCode = 401;
                throw new Error(dbCard[0].description);
            }

            // Check the month 
            if (month < 1 || month > 12) {
                statusCode = 400;
                throw new Error('Please check the month number');
            }

            // Check the cvv
            const cvvArr = cvv.split('');

            if (cvvArr.length < 3 || cvvArr.length > 4) {
                statusCode = 400;
                throw new Error('Please check your cvv number');
            }

            for (let i = 0; i < cvvArr.length; i++) {
                let num = Number(cvvArr[i]);

                if (String(num) === 'NaN') {
                    statusCode = 400;
                    throw new Error('Please check your cvv number');
                }
            }


            const id = idGenerator();
            const newPlanSign = {
                id,
                cardNum,
                month,
                year,
                cvv,
                ownerName
            };

            res.status(statusCode).send({message: 'Plan signed successfully', info: newPlanSign});
        } catch (error) {
            if (statusCode === 200) {
                statusCode = 500;
                res.status(statusCode).send({message: 'Internal server error, please try again'})
            }
            else {
                res.status(statusCode).send({message: error.message});
            }
        }
    });

    server.get('*', (req, res) => {
        return handle(req, res);
    });
    server.listen(port, (err) => {
        if (err) throw err
        console.log(`Server running on http://localhost:${port}`);
    })
}).catch((err) => {
    console.log(err);
    process.exit(1);
})